File: src/main/java/com/kyfstore/mcversionrenamer/MCVersionRenamerClient.java
package com.kyfstore.mcversionrenamer;

import com.kyfstore.mcversionrenamer.data.MCVersionPublicData;
import com.kyfstore.mcversionrenamer.event.KeyInputHandler;
import com.kyfstore.mcversionrenamer.libapi.core.version.VersionCheckerApi;
import com.kyfstore.mcversionrenamer.rewrites.MCVersionRenamerMinecraftVersionClass;
import com.kyfstore.mcversionrenamer.rewrites.MCVersionRenamerMinecraftGameVersion;
import net.fabricmc.api.ClientModInitializer;
import net.fabricmc.api.EnvType;
import net.fabricmc.api.Environment;
import net.fabricmc.fabric.api.client.event.lifecycle.v1.ClientTickEvents;
import net.fabricmc.fabric.api.client.screen.v1.ScreenEvents;
import net.minecraft.client.MinecraftClient;
import net.minecraft.client.gui.screen.TitleScreen;

@Environment(EnvType.CLIENT)
public class MCVersionRenamerClient implements ClientModInitializer {

    private static final VersionCheckerApi versionChecker = new VersionCheckerApi();

    private static boolean hasCheckedVersion = false;

    @Override
    public void onInitializeClient() {
        MCVersionRenamerMinecraftGameVersion versionClass = MCVersionRenamerMinecraftVersionClass.create();

        publicVersionClass = versionClass;

        KeyInputHandler.register();

        versionChecker.onEnable(this);

        ScreenEvents.AFTER_INIT.register((minecraftClient, screen, i, i1) -> {
            if (!MCVersionPublicData.fancyMenuIsLoaded) setClientWindowName(MCVersionRenamer.CONFIG.versionTextSettings.titleText());
        });

        ClientTickEvents.END_CLIENT_TICK.register(client -> {
            publicClient = client;
            MCVersionPublicData.versionText = MCVersionRenamer.CONFIG.versionTextSettings.versionText();
            MCVersionPublicData.titleText = MCVersionRenamer.CONFIG.versionTextSettings.titleText();
            MCVersionPublicData.f3Text = MCVersionRenamer.CONFIG.versionTextSettings.f3Text();
            if (client != null && client.getWindow() != null) {
                client.getWindow().setTitle(versionClass.getName());
                if (client.currentScreen instanceof TitleScreen && !hasCheckedVersion) {
                    hasCheckedVersion = true;
                    versionChecker.checkVersion(client);
                }
            }
        });
    }

    public static MinecraftClient publicClient;
    public static MCVersionRenamerMinecraftGameVersion publicVersionClass;

    public static void setClientWindowName(String newTitle) {
        if (publicVersionClass != null) {
            publicVersionClass.setName(newTitle);
        }
    }

    public static void toggleButtonVisibility() {
        MCVersionPublicData.customButtonIsVisible = !MCVersionPublicData.customButtonIsVisible;
    }
}

File: src/main/kotlin/com/kyfstore/mcversionrenamer/MCVersionRenamerClient.kt
package com.kyfstore.mcversionrenamer

import com.kyfstore.mcversionrenamer.event.KeyInputHandler
import com.kyfstore.mcversionrenamer.rewrites.MCVersionRenamerMinecraftVersionClass
import net.fabricmc.api.ClientModInitializer
import net.fabricmc.fabric.api.client.event.lifecycle.v1.ClientTickEvents
import com.kyfstore.mcversionrenamer.rewrites.MCVersionRenamerMinecraftGameVersion
import net.minecraft.client.MinecraftClient

class MCVersionRenamerClient : ClientModInitializer {

    override fun onInitializeClient() {
        val versionClass: MCVersionRenamerMinecraftGameVersion = MCVersionRenamerMinecraftVersionClass.create()

        publicVersionClass = versionClass

        KeyInputHandler.register()

        ClientTickEvents.END_CLIENT_TICK.register(ClientTickEvents.EndTick { client: MinecraftClient? ->
            publicClient = client
            if (client != null && client.window != null) {
                client.window.setTitle(versionClass.name)
            }
        })
    }

    companion object {
        @JvmStatic
        private var publicClient: MinecraftClient? = null

        @JvmStatic
        private var publicVersionClass: MCVersionRenamerMinecraftGameVersion? = null

        @JvmStatic
        fun setClientWindowName(newTitle: String) {
            publicVersionClass?.name = newTitle
        }
    }
}
